lib/utils/fanoutbuffer/buffer.go

```diff
@@ -0,0 +1,431 @@
+// attempt to read.
+type Buffer[T any] struct {
+	rw sync.RWMutex
+
+	// cfg holds buffer configuration.
+	cfg Config
+
+	// notify is closed and replaced each time an item is added to wake any cursors
+	// that are fully caught up.
+	notify chan struct{}
+
+	// closed indicates that the buffer has been closed.
+	closed bool
+
+	// cursors is the total number of cursors that exist. used to determine the number of cursors expected
+	// to observe each item.
+	cursors uint64
+
+	// pos is the absolute position of the oldest element in the ring, meaning that pos%Capacity is the index
+	// of that element.
+	pos uint64
+
+	// len is the total number of elements in the ring (i.e. the index of the newest element is at `(pos + len - 1) % Capacity`).
+	len uint64
+
+	// ring is the fixed-size circular buffer that serves as our primary storage location.
+	ring []entry[T]
+
+	// overflowStart is the time at which we began needing to overflow items.
+	overflowStart time.Time
+
+	// overflow is a variable length buffer of items that still need to be observed.
+	overflow []overflowEntry[T]
+}
+
+// entry represents an item in a buffer
+type entry[T any] struct {
+	item T
+	// wait is the number of cursors that still need to observe this item. each cursor
+	// decrements this value when it observes the item. items with wait==0 are cleaned
+	// up. Note that this atomic is only safe to decrement while read lock is held. Entries
+	// may be moved while write lock is held.
+	wait atomic.Uint64
+}
+
+type overflowEntry[T any] struct {
+	entry[T]
+	expires time.Time
+}
+
+// NewBuffer creates a new fanout buffer instance.
+func NewBuffer[T any](cfg Config) *Buffer[T] {
+	cfg.SetDefaults()
+	return &Buffer[T]{
+		cfg:    cfg,
+		notify: make(chan struct{}),
+		ring:   make([]entry[T], cfg.Capacity),
+	}
+}
+
+// Close permanently closes the fanout buffer. Note that all cursors are terminated immediately
+// and may therefore miss items that had not been read at the time Close is called.
+func (b *Buffer[T]) Close() {
+	b.write(func() {
+		b.closed = true
+		close(b.notify)
+	})
+}
+
+// NewCursor gets a new cursor into the buffer. Cursors *must* be closed as soon as they are
+// no longer being read from. Stale cursors may cause performance degredation.
+func (b *Buffer[T]) NewCursor() *Cursor[T] {
+	var c uint64
+	b.write(func() {
+		if b.closed {
+			return
+		}
+		c = b.pos + b.len
+		b.cursors++
+	})
+
+	cursor := &Cursor[T]{
+		buf:    b,
+		cursor: c,
+	}
+	runtime.SetFinalizer(cursor, finalizeCursor[T])
+	return cursor
+}
+
+// Append appends to the buffer and wakes all dormant cursors.
+func (b *Buffer[T]) Append(items ...T) {
+	var notify chan struct{}
+
+	b.write(func() {
+		if b.closed {
+			panic("Append called on closed fanout buffer")
+		}
+
+		// free up slots that are no longer needed
+		b.cleanupSlots()
+
+		for _, item := range items {
+			// ensure that there is at least one free slot in ring
+			b.ensureSlot()
```

Comment:
Might be better to ensure however many slots we need at once.
```suggestion
		// ensure that there is at least one free slot in ring
		b.ensureSlots(len(items))
		for _, item := range items {
```

Reply:
I tried doing this in an earlier iteration.  Performance improvements were minimal, and handling the case where `items` is larger than the ring buffer was messy, so I ended up reverting in favor of the simpler logic.